### PR TITLE
Cache parallel run quota

### PR DESCRIPTION
**Pull request by candicegjing**
This PR caches `user_info` to reduce database calls in `get_user_parallel_run_quota_left`:
The following status is collected using 100 staged jobs in staging and `get_user_parallel_run_quota_left` check is forced to run. In the real production environment, the situation might be different. But basically, this change will reduce around 40% of the time in `get_user_parallel_run_quota_left`, especially when a user who submitted a large number of jobs.
```
Sun Jan 26 18:57:48 2020    _schedule_run_bundles_on_workers.profile

         534465 function calls (466277 primitive calls) in 0.861 seconds

   Ordered by: cumulative time
   List reduced from 302 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    0.862    0.862 /opt/codalab-worksheets/codalab/server/bundle_manager.py:310(_schedule_run_bundles_on_workers)
      101    0.004    0.000    0.717    0.007 /opt/codalab-worksheets/codalab/model/bundle_model.py:2334(get_user_parallel_run_quota_left)
      101    0.002    0.000    0.289    0.003 /opt/codalab-worksheets/codalab/model/bundle_model.py:2283(get_user_info)
      203    0.001    0.000    0.258    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
      203    0.000    0.000    0.257    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
      203    0.000    0.000    0.256    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:322(_execute_on_connection)
      203    0.001    0.000    0.256    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:975(_execute_clauseelement)
      203    0.001    0.000    0.200    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:93(__exit__)
      203    0.002    0.000    0.185    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1061(_execute_context)
      203    0.000    0.000    0.160    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py:449(do_execute)
      203    0.002    0.000    0.160    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:163(execute)
      203    0.001    0.000    0.156    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:301(_query)
      203    0.148    0.001    0.148    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/connections.py:213(query)
      203    0.001    0.000    0.142    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:84(begin)
      406    0.002    0.000    0.140    0.000 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:52(contextual_connect)
      202    0.001    0.000    0.139    0.001 /opt/codalab-worksheets/codalab/server/worker_info_accessor.py:28(user_owned_workers)
      404    0.000    0.000    0.139    0.000 /opt/codalab-worksheets/codalab/server/worker_info_accessor.py:33(<genexpr>)
63428/202    0.059    0.000    0.138    0.001 /usr/lib/python3.6/copy.py:132(deepcopy)
      203    0.000    0.000    0.138    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/pool.py:490(<lambda>)
      202    0.002    0.000    0.137    0.001 /usr/lib/python3.6/copy.py:236(_deepcopy_dict)
```
After this change, the profiling resluts are:
```
Sun Jan 26 19:35:57 2020    _schedule_run_bundles_on_workers.profile

         69734 function calls (67273 primitive calls) in 0.384 seconds

   Ordered by: cumulative time
   List reduced from 280 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    0.384    0.384 /opt/codalab-worksheets/codalab/server/bundle_manager.py:310(_schedule_run_bundles_on_workers)
      102    0.003    0.000    0.377    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2334(get_user_parallel_run_quota_left)
      103    0.001    0.000    0.143    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:93(__exit__)
      103    0.000    0.000    0.131    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
      103    0.000    0.000    0.130    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
      103    0.000    0.000    0.130    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:322(_execute_on_connection)
      103    0.001    0.000    0.129    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:975(_execute_clauseelement)
      103    0.001    0.000    0.096    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1061(_execute_context)
      103    0.000    0.000    0.086    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py:449(do_execute)
      103    0.001    0.000    0.085    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:163(execute)
      103    0.001    0.000    0.083    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:301(_query)
      103    0.080    0.001    0.080    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/connections.py:213(query)
      103    0.000    0.000    0.076    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:84(begin)
      206    0.001    0.000    0.074    0.000 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:52(contextual_connect)
      103    0.000    0.000    0.072    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/pool.py:490(<lambda>)
      103    0.001    0.000    0.072    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/pool.py:615(_finalize_fairy)
      103    0.000    0.000    0.071    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:105(commit)
      103    0.000    0.000    0.070    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1569(commit)
      103    0.000    0.000    0.070    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1603(_do_commit)
      103    0.001    0.000    0.070    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:679(_commit_impl)
```

### CODE REVIEW

File: codalab/server/bundle_manager.py [hunk 1]
```diff
@@ -307,7 +307,7 @@ def _bring_offline_stuck_running_bundles(self, workers):
                 logger.info('Bringing bundle offline %s: %s', bundle.uuid, failure_message)
                 self._model.transition_bundle_worker_offline(bundle)
 
-    def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run):
+    def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run, user_info_cache):
```
Comment (percyliang) [inline codalab/server/bundle_manager.py:310]: Update docstring

File: codalab/server/bundle_manager.py [hunk 2]
```diff
@@ -619,10 +624,10 @@ def _schedule_run_bundles(self):
         self._restage_stuck_starting_bundles(workers)
         self._bring_offline_stuck_running_bundles(workers)
         self._acknowledge_recently_finished_bundles(workers)
-        staged_bundles_to_run = self._get_staged_bundles_to_run()
+        staged_bundles_to_run, user_info_cache = self._get_staged_bundles_to_run()
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: To make the function signatures more uniform and clean, I'd create `user_info_cache = {}` and pass it in to all the functions that need it; it's just a variable that will get accessed and updated as needed.

**Review by percyliang — CHANGES_REQUESTED** [review]

File: codalab/server/bundle_manager.py [hunk 3]
```diff
@@ -307,11 +307,14 @@ def _bring_offline_stuck_running_bundles(self, workers):
...
         """
-        Schedules STAGED bundles to run on the given workers. If user_owned is
-        True, then schedules on workers run by the owner of each bundle.
-        Otherwise, uses CodaLab-owned workers, which have user ID root_user_id.
+        Schedules STAGED bundles to run on the given workers. Always tries to schedule bundles to
+        run on workers that are owned by the owner of each bundle first. If there are no such qualified
+        private workers, uses CodaLab-owned workers, which have user ID root_user_id.
+        :param workers: a WorkerInfoAccessor object containing worker related information e.g. running uuid
+        :param staged_bundles_to_run: a list of tuples each contains a valid bundle and its bundle resources
+        :param user_info_cache: a dictionary storing the mapping from user id to user information
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Simplify: "dictionary mapping user id to user information"

File: codalab/server/bundle_manager.py [hunk 4]
```diff
@@ -651,18 +660,17 @@ def _check_resource_failure(
...
         return None
 
-    def _get_staged_bundles_to_run(self):
+    def _get_staged_bundles_to_run(self, user_info_cache):
         """
         Fails bundles that request more resources than available for the given user.
         Note: allow more resources than available on any worker because new
         workers might get spun up in response to the presence of this run.
-
+        :param user_info_cache: a dictionary structured as {user id : user information} to track
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Why not use the same doc string as the other instance?

**Review by percyliang — APPROVED** [review]
Great!

File: codalab/server/bundle_manager.py (hunk 4 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py]: Yeah, sure. That would be better I think

File: codalab/server/bundle_manager.py (hunk 3 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py]: Updated

**Pull request merged**